server/keyspace/keyspace.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keyspace
+
+import (
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"github.com/tikv/pd/pkg/syncutil"
+	"github.com/tikv/pd/server/id"
+	"github.com/tikv/pd/server/storage/endpoint"
+)
+
+const (
+	spaceIDMin = uint32(1)       // 1 is the minimum value of spaceID, 0 is reserved.
+	spaceIDMax = ^uint32(0) >> 8 // 16777215 (Uint24Max) is the maximum value of spaceID.
+)
+
+var (
+	errKeyspaceArchived   = errors.New("Keyspace already archived")
+	ErrKeyspaceNotFound   = errors.New("Keyspace does not exist")
+	ErrKeyspaceNameExists = errors.New("Keyspace name already in use")
+	errIllegalID          = errors.New("Cannot create keyspace with that ID")
+	errIllegalName        = errors.New("Cannot create keyspace with that name")
+)
+
+// Manager manages keyspace related data.
+// It validates requests and provides concurrency control.
+type Manager struct {
+	// idLock guards keyspace name to id lookup entries.
+	idLock syncutil.Mutex
+	// metaLock guards keyspace meta.
+	metaLock syncutil.Mutex
+	// idAllocator allocates keyspace id.
+	idAllocator id.Allocator
+	// store is the storage for keyspace related information.
+	store endpoint.KeyspaceStorage
+}
+
+// CreateKeyspaceRequest represents necessary arguments to create a keyspace.
+type CreateKeyspaceRequest struct {
+	Name          string
+	InitialConfig map[string]string
+	Now           time.Time
+}
+
+// UpdateKeyspaceRequest represents necessary arguments to update keyspace.
+type UpdateKeyspaceRequest struct {
+	Name string
+	// Whether to update the state of keyspace,
+	// if set to false, NewState and Now will be ignored.
+	UpdateState bool
+	NewState    keyspacepb.KeyspaceState
+	Now         time.Time
+	// New KVs to put in keyspace config.
+	ToPut map[string]string
```

Comment:
Why there is extra 'To' prefix?

Reply:
Mostly to disambiguate with go's `delete` built-in, which was used in this package.